### PR TITLE
PS-4439: Port clustering secondary index patch to 8.0

### PR DESCRIPTION
**Pull request by satya-bodapati**
Summary:
--------
Support CLUSTERING KEY type of TokuDB with new Data Dictionary

A new options property "clustering_key" is used to determine the
CLUSTERING KEY type. This is stored as part of dd::Index::options

KEY* flag HA_CLUSTERING is set based on clustering key property of
dd::Index::options (fill_index_from_dd())

Testing which are failing due to collation changes, "row" count
mismatch of explain output (wrong column masked) are rerecorded.

### CODE REVIEW

File: sql/dd_table_share.cc [hunk 1]
```diff
@@ -1187,6 +1187,16 @@ static bool fill_index_from_dd(THD *thd, TABLE_SHARE *share,
       break;
   }
 
+  /* Check if this index is of clustering key type. Used by TokuDB */
+  const dd::Properties *index_options = &idx_obj->options();
```
Comment (laurynas-biveinis) [inline sql/dd_table_share.cc]: nit: why not a reference instead of pointer?

**Review by laurynas-biveinis — APPROVED** [review]

File: sql/dd_table_share.cc (hunk 1 above)
Comment (satya-bodapati) [reply sql/dd_table_share.cc]: can be written that way too. Updated the patch

File: mysql-test/t/tokudb_clustering_key_grammar.test [hunk 2]
```diff
@@ -51,7 +51,8 @@ ALTER TABLE t1 ADD CONSTRAINT c1 UNIQUE CLUSTERING INDEX b (b);
 ALTER TABLE t1 ADD CONSTRAINT c1 CLUSTERING UNIQUE INDEX b (b);
 --error ER_PARSE_ERROR
 ALTER TABLE t1 ADD CONSTRAINT c1 UNIQUE UNIQUE INDEX b (b);
---error ER_PARSE_ERROR
+#--error ER_PARSE_ERROR
```
Comment (george-lorch) [inline mysql-test/t/tokudb_clustering_key_grammar.test]: Just remove, no need to keep commented error expectation

**Review by george-lorch — APPROVED** [review]
Loos good, all test result changes are perfectly in line with what is known to have changed in 8.0.

**Pull request merged**